tensornetwork/backends/jax/jax_backend.py

```diff
@@ -222,6 +226,115 @@ def cmplx_random_uniform(complex_dtype, real_dtype):
+    import jax
+    import numpy as np
+    for n in range(100):
+      def A(H,x):
+        return jax.np.dot(H,x)
+      H = jax.np.array(np.random.rand(10,10))
+      x = jax.np.array(np.random.rand(10,10))
+      res = eigsh_lanczos(A, [H],x) #jitting is triggerd at every step `n`
+    ```
+    
+    Args:
+      A: A (sparse) implementation of a linear operator.
+         Call signature of `A` is `res = A(*args, vector)`, where `vector`
+         can be an arbitrary `Tensor`, and `res.shape` has to be `vector.shape`.
+      arsg: A list of arguments to `A`.  `A` will be called as
+        `res = A(*args, initial_state)`.
+      initial_state: An initial vector for the Lanczos algorithm. If `None`,
+        a random initial `Tensor` is created using the `backend.randn` method
+      shape: The shape of the input-dimension of `A`.
+      dtype: The dtype of the input `A`. If both no `initial_state` is provided,
+        a random initial state with shape `shape` and dtype `dtype` is created.
+      num_krylov_vecs: The number of iterations (number of krylov vectors).
+      numeig: The nummber of eigenvector-eigenvalue pairs to be computed.
+        If `numeig > 1`, `reorthogonalize` has to be `True`.
+      tol: The desired precision of the eigenvalus. Uses
+        `np.linalg.norm(eigvalsnew[0:numeig] - eigvalsold[0:numeig]) < tol`
+        as stopping criterion between two diagonalization steps of the
+        tridiagonal operator.
+      delta: Stopping criterion for Lanczos iteration.
+        If a Krylov vector :math: `x_n` has an L2 norm
+        :math:`\\lVert x_n\\rVert < delta`, the iteration
+        is stopped. It means that an (approximate) invariant subspace has
+        been found.
+      ndiag: The tridiagonal Operator is diagonalized every `ndiag` iterations
+        to check convergence.
+      reorthogonalize: If `True`, Krylov vectors are kept orthogonal by
+        explicit orthogonalization (more costly than `reorthogonalize=False`)
+    Returns:
+      (eigvals, eigvecs)
+       eigvals: A list of `numeig` lowest eigenvalues
+       eigvecs: A list of `numeig` lowest eigenvectors
+    """
+
+    if num_krylov_vecs < numeig:
+      raise ValueError('`num_krylov_vecs` >= `numeig` required!')
+
+    if numeig > 1 and not reorthogonalize:
+      raise ValueError(
+          "Got numeig = {} > 1 and `reorthogonalize = False`. "
+          "Use `reorthogonalize=True` for `numeig > 1`".format(numeig))
+    if initial_state is None:
+      if (shape is None) or (dtype is None):
+        raise ValueError("if no `initial_state` is passed, then `shape` and"
+                         "`dtype` have to be provided")
+      initial_state = self.randn(shape, dtype)
+
+    if not isinstance(initial_state, jnp.ndarray):
+      raise TypeError("Expected a `jax.array`. Got {}".format(
+          type(initial_state)))
+    if hasattr(self, '_A'):
+      #pylint: disable=access-member-before-definition
+      if self._A is not A:
```

Comment:
Why are you setting the passed function as a attribute to the backend? Seems very sketchy...

Reply:
Yeah, it is kinda sketchy... The reason is that I want to avoid retracing. 
`_generate_jitted_eighs_lanczos` jits the function `jax_lanczos` in jitted_functions.py. The first argument to `jax_lanczos` is `matvec`, a  callable that does the matrix-vector product. that function `matvec` has to be wrapped into `jax.tree_util.Partial`, otherwise `jax.jit` will complain. Now, `matvec` is typically passed by the user (name `A` there), but I'd like to avoid having users wrapping their `matvecs` into `jax.tree_util.Partial` because that breaks backend independence of their code. Thus I wrap it in the backend function `eigsh_lanczos`. Unfortunately, if I just wrap it at every call to `eigs_lanczos` that would trigger jitting of `jax_lanczos` at every  call, something I want to avoid. So at the first call to `eigsh_lanczos` I wrape `A` into `jax.tree_util.Partial` and store both the passed `A` and the wrapped `A`. If the user calls `eigsh_lanczos` again with the same `A`  I can check if I already wrapped `A`, and I can just pass the previously wrapped `A` (stored in `self._APartial`) to `jax_lanczos`, thus avoiding rejitting. Couldn't really find a better way to accomplish that

Reply:
Can we do this instead?

Let's have a dictonary for this module called `_CACHED_FUNCTIONS = {}` and use that to store the methods. That way we avoid adding an unnecessary attribute to the backend AND we can cache more than one method at a time.

Reply:
that's actually a great idea, let me do that